libs/go/sia/gcp/functions/gcf_get_sia_cert.go

```diff
@@ -0,0 +1,263 @@
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"github.com/AthenZ/athenz/clients/go/zts"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// GcfGetSiaCerts this method can be called from within a GCF (Google Cloud Function) -
+//
+//	to get an Athenz certificate from ZTS.
+//
+// This file should usually be copied without changes into the GCF source-code.
+//
+// See https://cloud.google.com/functions/docs/writing/write-http-functions#http-example-go
+func GcfGetSiaCerts(
+	athenzDomain string,
+	athenzService string,
+	gcpProjectId string,
+	athenzProvider string,
+	ztsUrl string,
+	certDomain string,
+	optionalSubjectFields CsrSubjectFields,
+) (*SiaCertData, error) {
+
+	athenzDomain = strings.ToLower(athenzDomain)
+	athenzService = strings.ToLower(athenzService)
+	athenzProvider = strings.ToLower(athenzProvider)
+
+	// Get an identity-document for this GCF from GCP.
+	attestationData, err := getGcpFunctionAttestationData(ztsUrl)
+	if err != nil {
+		return nil, err
+	}
+	//log.Printf("GCP Attestation Data: %s", attestationData) // commented out - sensitive info
+
+	// Create a private-key.
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the private-key to PEM format.
+	privateKeyDer := x509.MarshalPKCS1PrivateKey(privateKey)
+	privateKeyBlock := pem.Block{Type: "RSA PRIVATE KEY", Bytes: privateKeyDer}
+	privateKeyPem := pem.EncodeToMemory(&privateKeyBlock)
+	//log.Printf("Private Key:\n%s", privateKeyPem) // commented out - sensitive info
+
+	// Create a CSR (and a private-key).
+	csr, err := generateCsr(
+		privateKey,
+		athenzDomain+"."+athenzService,
+		optionalSubjectFields,
+		[]string{},
+		[]string{
+			athenzService + "." + strings.Replace(athenzDomain, ".", "-", -1) + "." + certDomain,
+		},
+		[]string{
+			"spiffe://" + athenzDomain + "/sa/" + athenzService,
+			"athenz://instanceid/" + athenzProvider + "/gcp-function-" + gcpProjectId,
+		})
+
+	// Encode the CSR to PEM.
+	var csrPemBuffer bytes.Buffer
+	err = pem.Encode(&csrPemBuffer, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csr})
+	if err != nil {
+		return nil, fmt.Errorf("cannot encode CSR to PEM: %v", err)
+	}
+	csrPem := csrPemBuffer.String()
+	//log.Printf("CSR to send to ZTS:\n%s", csrPem) // commented out - sensitive info
+
+	// Send CSR to ZTS.
+	siaCertData, err := getCredsFromZts(
+		ztsUrl,
+		zts.InstanceRegisterInformation{
+			Domain:          zts.DomainName(athenzDomain),
+			Service:         zts.SimpleName(athenzService),
+			Provider:        zts.ServiceName(athenzProvider),
+			AttestationData: attestationData,
+			Csr:             csrPem,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	siaCertData.privateKey = privateKey
+	siaCertData.privateKeyPem = string(privateKeyPem)
+	return siaCertData, nil
+}
+
+// SiaCertData response of GcfGetSiaCerts()
+type SiaCertData struct {
```

Comment:
No need to define a new struct. This is already available as part of our RDL

import 	"github.com/AthenZ/athenz/clients/go/zts"

and generate and return zts.InstanceIdentity object

Reply:
Thing is, zts.InstanceIdentity is JSON-able: it only contain PEM stuff.
My SiaCertData also contains rsa.PrivateKey and x509.Certificate objects - so usage is simpler.
Also, zts.InstanceIdentitycontains many properties that I will not fill-in, and it would be confusing to users.
I would not make this change, but will yield to your judgment...
Are you sure you want to go that way?